lib/pairmotron/pair_builder.ex

```diff
@@ -0,0 +1,94 @@
+defmodule Determination do
+  defstruct dead_pairs: [], remaining_pairs: [], available_users: []
+end
+
+defmodule Pairmotron.PairBuilder do
+  @moduledoc """
+  Provides ways of determining what has changed between an existing set of %UserPair{}s and a new
+  set of %Users{}s that need to be paired.
+  %UserPair{}s are, for the intent of this module,  records of how users were paired previously.
+
+  This module will only spell out the difference  between how users have already been paired and
+  what users are now able to be paired.
+  %Pair{}s and %UserPair{}s may either not change or be deleted.
+  %Users{}s will be free to find a new pair.
+  """
+
+
+  @doc """
+  Find the dead pairs, remaining user pairs, and available users given the previous %UserPair{} records
+  and the new list of %User{}s.
+  """
+  def determify(user_pairs, users) do
+    dead_user_pairs = find_dead_user_pairs(user_pairs, users)
+    %Determination{
+      dead_pairs: dead_user_pairs |> find_dead_pairs,
+      remaining_pairs: dead_user_pairs |> find_remaining_user_pairs(user_pairs) |> unique_pairs,
+      available_users: dead_user_pairs |> find_users_to_pair(user_pairs, users)
+    }
+  end
```

Comment:
Why was the decision made to make the arguments `user_pairs` and `users` rather than `pairs` and `users`? Correct me if I'm wrong, but is there ever a case where you want to only delete some of the `UserPair`s associated with a pair? If not, couldn't we just delete the pairs that are no longer valid and rely on the database constraints to take care of the `UserPair` records that need to be deleted?

I feel like this would remove a lot of the logic you had to use to relate `Pair`, `UserPair` and `User` records in various ways, as you can reference the `:users` association on the pairs directly to free up users. I believe that this could be a nice performance boost too, but it's hard to say without doing it.

I wouldn't say change it for this PR, because this is working well, but that's definitely a candidate for a possible refactor down the line.